docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,804 @@
+
+```go
+// CryptoProviderPV is the implementation of PrivValidator using CryptoProvider's methods
+type CryptoProviderPV struct {
+    provider CryptoProvider
+    pubKey   PubKey
+}
+
+// NewCryptoProviderPV creates a new instance of CryptoProviderPV
+func NewCryptoProviderPV(provider CryptoProvider, pk PubKey) (*CryptoProviderPV, error) {
+    return &CryptoProviderPV{provider: provider, pubKey: pubKey}, nil
+}
+
+// GetPubKey returns the public key of the validator
+func (pv *CryptoProviderPV) GetPubKey() (PubKey, error) {
+    return pv.pubKey, nil
+}
+
+// SignVote signs a canonical representation of the vote. If signExtension is true, it also signs the vote extension.
+func (pv *CryptoProviderPV) SignVote(chainID string, vote *Vote, signExtension bool) error {
+    signer := pv.provider.GetSigner()
+
+    // code for getting voteBytes goes here
+    // voteBytes := ...
+
+    // The underlying signer needs these parameters so we pass them through SignerOptions
+    options := SignerOptions{
+        "chainID": chainID,
+        "vote":    vote,
+    }
+ 
+    sig, _ := signer.Sign(voteBytes, options)
+    vote.Signature = sig
+    return nil
+}
+
+// SignProposal signs a canonical representation of the proposal
+func (pv *CryptoProviderPV) SignProposal(chainID string, proposal *Proposal) error {
+    signer := pv.provider.GetSigner()
+
+    // code for getting proposalBytes goes here
+    // proposalBytes := ...
+
+    // The underlying signer needs these parameters so we pass them through SignerOptions
+    options := SignerOptions{
+        "chainID":  chainID,
+        "proposal": proposal,
+    }
+ 
+    sig, _ := signer.Sign(proposalBytes, options)
+    proposal.Signature = sig
+    return nil
+}
+
+// SignBytes signs an arbitrary array of bytes
+func (pv *CryptoProviderPV) SignBytes(bytes []byte) ([]byte, error) {
+    signer := pv.provider.GetSigner()
+    return signer.Sign(bytes, SignerOptions{})
+}
+
+```
+
+*Note:* Each provider (File, Socket, SignerClient, RetrySignerClient) will need to be modified to satisfy the `CryptoProvider` interface.
+
+
+#### Loading and storing implementations
+
+Alternatives:
+
+* Low impact / minimal dependencies: The corresponding CryptoProvider [Protobuf message](#protobuf-message-structure) can be directly stored on disk in a dedicated directory in an encoding format of choice (text, JSON).
+
+* Greater impact / better security: Use cosmos-sdk's [Keyring](https://github.com/cosmos/cosmos-sdk/blob/439f2f9d5b5884bc9df4b58d702555330549a898/crypto/keyring/keyring). to manage CryptoProviders along with its private keys. This specifically applies to the `FilePV` implementation, which could store its private keys in Keyring instead of a file in the filesystem. This approach will require decoupling the Keyring package from the cosmos-sdk to avoid having CometBFT import artifacts from that repository.
```

Comment:
While, in the long term, this might be the way to go (although I still think using a keyring at CometBFT level is overkill), I think this change will require changes in the basic way operators function today. I'd prefer to defer this for a further enhancement, once we have the the new design implemented, working, and no major pushback from users.

Reply:
Sure, this can be added to the list of possible future enhancements.